### PR TITLE
Refine catalog of partition & alias.

### PR DESCRIPTION
**Pull request by longjiquan**
Signed-off-by: longjiquan <jiquan.long@zilliz.com>
issue: #17412

### CODE REVIEW

File: internal/metastore/kv/kv_catalog.go [hunk 1]
```diff
@@ -226,22 +312,73 @@ func (kc *Catalog) CreateCredential(ctx context.Context, credential *model.Crede
...
+	collMeta, err := kc.loadCollection(ctx, collectionID, ts)
 	if err != nil {
 		log.Error("collection meta marshal fail", zap.String("key", collKey), zap.Error(err))
 		return nil, err
 	}
 
-	return model.UnmarshalCollectionModel(collMeta), nil
+	collection := model.UnmarshalCollectionModel(collMeta)
+
+	if partitionVersionAfter210(collMeta) {
```
Comment (jaime0815) [inline internal/metastore/kv/kv_catalog.go:365]: We' better  check whether need to process compatibility explicitly, it should return collection meta if the version is less than 210

File: internal/metastore/kv/kv_catalog.go [hunk 2]
```diff
@@ -306,45 +443,60 @@ func (kc *Catalog) DropCollection(ctx context.Context, collectionInfo *model.Col
...
-		log.Error("drop partition marshal fail", zap.String("key", k), zap.Error(err))
-		return err
+func dropPartition(collMeta *pb.CollectionInfo, partitionID typeutil.UniqueID) {
+	if collMeta == nil {
+		return
+	}
+
+	{
+		loc := -1
+		for idx, partition := range collMeta.GetPartitions() {
```
Comment (jaime0815) [inline internal/metastore/kv/kv_catalog.go:453]: revert partitions addition from collection and ignore its' compatibility

File: internal/metastore/kv/kv_catalog.go [hunk 3]
```diff
@@ -441,31 +585,73 @@ func (kc *Catalog) ListCollections(ctx context.Context, ts typeutil.Timestamp) (
...
+		colls[collMeta.Schema.Name] = collection
 	}
 
 	return colls, nil
 }
 
-func (kc *Catalog) ListAliases(ctx context.Context) ([]*model.Collection, error) {
-	_, values, err := kc.Snapshot.LoadWithPrefix(CollectionAliasMetaPrefix, 0)
+func (kc *Catalog) listAliasesBefore210(ctx context.Context, ts typeutil.Timestamp) ([]*model.Alias, error) {
+	log.Debug("enter before 210 ..........")
```
Comment (jaime0815) [inline internal/metastore/kv/kv_catalog.go:599]: remove this log  or add more detailed information.

File: internal/metastore/kv/kv_catalog.go [hunk 4]
```diff
@@ -441,31 +585,73 @@ func (kc *Catalog) ListCollections(ctx context.Context, ts typeutil.Timestamp) (
...
+			CollectionID: coll.GetID(),
+			CreatedTime:  0, // not accurate.
+		})
+	}
+	return aliases, nil
+}
 
-	var colls []*model.Collection
+func (kc *Catalog) listAliasesAfter210(ctx context.Context, ts typeutil.Timestamp) ([]*model.Alias, error) {
+	log.Debug("enter after 210 ..........")
```
Comment (jaime0815) [inline internal/metastore/kv/kv_catalog.go:622]: the same with the above

File: internal/metastore/model/collection.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ type Collection struct {
 	StartPositions       []*commonpb.KeyDataPair
 	CreateTime           uint64
 	ConsistencyLevel     commonpb.ConsistencyLevel
-	Aliases              []string
+	Aliases              []string          // TODO: deprecate this.
```
Comment (jaime0815) [inline internal/metastore/model/collection.go:25]: Maybe it should maintain within the collection model, we need a mapping from the collection id to all aliases of itself

File: internal/proto/etcd_meta.proto [hunk 6]
```diff
@@ -36,13 +36,20 @@ message CollectionInfo {
   int32 shards_num = 10;
   repeated common.KeyDataPair start_positions = 11;
   common.ConsistencyLevel consistency_level = 12;
-  repeated PartitionInfo partitions = 13;
+  repeated PartitionInfo partitions = 13; // deprecated.
```
Comment (jaime0815) [inline internal/proto/etcd_meta.proto:39]: remove it and don't consider compatibility issues.

File: internal/metastore/catalog.go [hunk 7]
```diff
@@ -17,19 +17,19 @@ type Catalog interface {
 	CollectionExists(ctx context.Context, collectionID typeutil.UniqueID, ts typeutil.Timestamp) bool
 	DropCollection(ctx context.Context, collectionInfo *model.Collection, ts typeutil.Timestamp) error
 
-	CreatePartition(ctx context.Context, coll *model.Collection, ts typeutil.Timestamp) error
-	DropPartition(ctx context.Context, collectionInfo *model.Collection, partitionID typeutil.UniqueID, ts typeutil.Timestamp) error
+	CreatePartitionV2(ctx context.Context, partition *model.Partition, ts typeutil.Timestamp) error
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/catalog.go:20]: OK, I'll modify these.

File: internal/metastore/kv/kv_catalog.go [hunk 8]
```diff
@@ -28,60 +28,151 @@ import (
...
 	Txn      kv.TxnKV
 	Snapshot kv.SnapShotKV
 }
 
+func buildCollectionKey(collectionID typeutil.UniqueID) string {
+	return fmt.Sprintf("%s/%d", CollectionMetaPrefix, collectionID)
+}
+
+func buildPartitionKey(collectionID, partitionID typeutil.UniqueID) string {
+	return fmt.Sprintf("%s/%d/%d", PartitionMetaPrefix, collectionID, partitionID)
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go:45]: OK, I'll fix this.

File: internal/metastore/kv/kv_catalog.go [hunk 9]
```diff
@@ -28,60 +28,151 @@ import (
...
+}
+
+func buildPartitionPrefix(collectionID typeutil.UniqueID) string {
+	return fmt.Sprintf("%s/%d", PartitionMetaPrefix, collectionID)
+}
+
+func buildFieldKey(collectionID typeutil.UniqueID, fieldID int64) string {
+	return fmt.Sprintf("%s/%d/%d", FieldMetaPrefix, collectionID, fieldID)
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go:53]: OK, I'll fix this.

File: internal/metastore/kv/kv_catalog.go [hunk 10]
```diff
@@ -28,60 +28,151 @@ import (
...
+		fieldInfo := model.MarshalFieldModel(field)
+		v, err := proto.Marshal(fieldInfo)
+		if err != nil {
+			return err
+		}
+		kvs[k] = string(v)
 	}
 
-	return nil
+	return kc.Snapshot.MultiSave(kvs, ts)
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go:99]: We should create snapshot for these meta, however `txn` won't.

File: internal/metastore/kv/kv_catalog.go [hunk 11]
```diff
@@ -28,60 +28,151 @@ import (
...
+}
+
+func partitionVersionAfter210(collMeta *pb.CollectionInfo) bool {
+	return len(collMeta.GetPartitionIDs()) <= 0 &&
+		len(collMeta.GetPartitionNames()) <= 0 &&
+		len(collMeta.GetPartitionCreatedTimestamps()) <= 0 &&
+		len(collMeta.GetPartitions()) <= 0
+}
+
+func partitionExist(collMeta *pb.CollectionInfo, partitionID typeutil.UniqueID) bool {
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go:122]: OK.

File: internal/metastore/kv/kv_catalog.go [hunk 12]
```diff
@@ -28,60 +28,151 @@ import (
...
+		k := buildPartitionKey(partition.CollectionID, partition.PartitionID)
+		partitionInfo := model.MarshalPartitionModel(partition)
+		v, err := proto.Marshal(partitionInfo)
+		if err != nil {
+			return err
+		}
+		return kc.Snapshot.Save(k, string(v), ts)
+	}
+
+	if partitionExist(collMeta, partition.PartitionID) {
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go:157]: OK.

File: internal/metastore/kv/kv_catalog.go (hunk 1 above)
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go]: OK, fixed.

File: internal/metastore/kv/kv_catalog.go (hunk 2 above)
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go]: removed.

File: internal/metastore/kv/kv_catalog.go (hunk 4 above)
Comment (longjiquan) [reply internal/metastore/kv/kv_catalog.go]: removed.

File: internal/proto/etcd_meta.proto (hunk 6 above)
Comment (longjiquan) [reply internal/proto/etcd_meta.proto]: removed.

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

File: internal/metastore/kv/rootcorod_constant.go [hunk 13]
```diff
@@ -7,6 +7,10 @@ const (
 	// CollectionMetaPrefix prefix for collection meta
 	CollectionMetaPrefix = ComponentPrefix + "/collection"
 
+	PartitionMetaPrefix = ComponentPrefix + "/partitions"
+	AliasMetaPrefix     = ComponentPrefix + "/aliases"
+	FieldMetaPrefix     = ComponentPrefix + "/fields"
+
```
Comment (aakejiang) [inline internal/metastore/kv/rootcorod_constant.go:13]: typo for  `rootcorod_constant.go`

File: internal/metastore/kv/kv_catalog.go [hunk 14]
```diff
@@ -28,60 +28,185 @@ import (
...
 	}
 
-	kvs := map[string]string{k1: string(v1)}
-	err = kc.Snapshot.MultiSave(kvs, ts)
+	collMeta := &pb.CollectionInfo{}
+	err = proto.Unmarshal([]byte(collVal), collMeta)
+	return collMeta, err
+}
+
+func partitionVersionAfter210(collMeta *pb.CollectionInfo) bool {
```
Comment (czs007) [inline internal/metastore/kv/kv_catalog.go:160]: functions return bool better to named with "is" prefix.
you can change this in next pr.

File: internal/metastore/kv/kv_catalog.go [hunk 15]
```diff
@@ -441,31 +604,71 @@ func (kc *Catalog) ListCollections(ctx context.Context, ts typeutil.Timestamp) (
...
+	for _, value := range values {
+		coll := &pb.CollectionInfo{}
+		err := proto.Unmarshal([]byte(value), coll)
+		if err != nil {
+			return nil, err
+		}
+		aliases = append(aliases, &model.Alias{
+			Name:         coll.GetSchema().GetName(),
+			CollectionID: coll.GetID(),
+			CreatedTime:  0, // not accurate.
```
Comment (czs007) [inline internal/metastore/kv/kv_catalog.go:633]: Better to change this to collection's createTime.
you can change this in next pr.

**Pull request merged**